src/cmd/services/m3coordinator/downsample/downsample_id.go

```diff
@@ -0,0 +1,330 @@
+	tagDecoderPool serialize.TagDecoderPool,
+	opts pool.ObjectPoolOptions,
+) *encodedTagsIteratorPool {
+	return &encodedTagsIteratorPool{
+		tagDecoderPool: tagDecoderPool,
+		pool:           pool.NewObjectPool(opts),
+	}
+}
+
+func (p *encodedTagsIteratorPool) Init() {
+	p.tagDecoderPool.Init()
+	p.pool.Init(func() interface{} {
+		return newEncodedTagsIterator(p.tagDecoderPool.Get(), p)
+	})
+}
+
+func (p *encodedTagsIteratorPool) Get() *encodedTagsIterator {
+	return p.pool.Get().(*encodedTagsIterator)
+}
+
+func (p *encodedTagsIteratorPool) Put(v *encodedTagsIterator) {
+	p.pool.Put(v)
+}
+
+func isRollupID(
+	tags []byte,
+	iteratorPool *encodedTagsIteratorPool,
+) bool {
+	iter := iteratorPool.Get()
+	iter.Reset(tags)
+
+	for iter.Next() {
+		name, value := iter.Current()
+		if bytes.Equal(name, rollupTagName) && bytes.Equal(value, rollupTagValue) {
+			return true
+		}
+	}
+	iter.Close()
+
+	return false
+}
+
+type rollupIDProvider struct {
+	index          int
+	tagPairs       []id.TagPair
+	rollupTagIndex int
+
+	tagEncoder serialize.TagEncoder
+	pool       *rollupIDProviderPool
+}
+
+func newRollupIDProvider(
+	tagEncoder serialize.TagEncoder,
+	pool *rollupIDProviderPool,
+) *rollupIDProvider {
+	return &rollupIDProvider{
+		tagEncoder: tagEncoder,
+		pool:       pool,
+	}
+}
+
+func (p *rollupIDProvider) provide(
+	tagPairs []id.TagPair,
+) ([]byte, error) {
+	p.reset(tagPairs)
+	p.tagEncoder.Reset()
+	if err := p.tagEncoder.Encode(p); err != nil {
+		return nil, err
+	}
+	data, ok := p.tagEncoder.Data()
+	if !ok {
+		return nil, fmt.Errorf("unable to access encoded tags: ok=%v", ok)
+	}
+	// Need to return a copy
+	return append([]byte(nil), data.Bytes()...), nil
+}
+
+func (p *rollupIDProvider) reset(
+	tagPairs []id.TagPair,
+) {
+	p.index = -1
+	p.tagPairs = tagPairs
+	p.rollupTagIndex = -1
+	for idx, pair := range tagPairs {
+		if bytes.Compare(rollupTagName, pair.Name) < 0 {
+			p.rollupTagIndex = idx
+			break
+		}
+	}
+	if p.rollupTagIndex == -1 {
+		p.rollupTagIndex = len(p.tagPairs)
+	}
+}
+
+func (p *rollupIDProvider) length() int {
+	return len(p.tagPairs) + 1
+}
+
+func (p *rollupIDProvider) finalize() {
+	if p.pool != nil {
+		p.pool.Put(p)
+	}
+}
+
+func (p *rollupIDProvider) Next() bool {
+	p.index++
+	return p.index < p.length()
+}
+
+func (p *rollupIDProvider) Current() ident.Tag {
+	idx := p.index
+	if idx == p.rollupTagIndex {
+		return ident.Tag{
+			Name:  ident.BytesID(rollupTagName),
+			Value: ident.BytesID(rollupTagValue),
+		}
+	}
+
+	if idx > p.rollupTagIndex {
+		// Effective index is subtracted by 1
+		idx--
+	}
+
+	return ident.Tag{
+		Name:  ident.BytesID(p.tagPairs[idx].Name),
+		Value: ident.BytesID(p.tagPairs[idx].Value),
+	}
+}
+
+func (p *rollupIDProvider) Err() error {
+	return nil
+}
+
+func (p *rollupIDProvider) Close() {
+	// No-op
+}
+
+func (p *rollupIDProvider) Remaining() int {
+	return p.length() - p.index - 1
+}
+
+func (p *rollupIDProvider) Duplicate() ident.TagIterator {
```

Comment:
This only does a shallow copy (e.g., the tag encoder is shared between itself and the clone). Is that ok?

Reply:
I only made this to fulfill the interface, in practice it won't be duplicated.  I can make a deep copy though just to be careful.